chain/ethereum/src/network.rs

```diff
@@ -64,17 +91,76 @@ impl EthereumNetworkAdapters {
     pub fn cheapest(&self) -> Option<Arc<EthereumAdapter>> {
         // EthereumAdapters are sorted by their NodeCapabilities when the EthereumNetworks
         // struct is instantiated so they do not need to be sorted here
-        self.adapters
+        self.full_adapters
             .first()
             .map(|ethereum_network_adapter| ethereum_network_adapter.adapter.clone())
     }
 
+    pub fn all_cheapest_call_only_with(
+        &self,
+        required_capabilities: &NodeCapabilities,
+    ) -> impl Iterator<Item = Arc<EthereumAdapter>> + '_ {
+        all_cheapest_with(&self.full_adapters, required_capabilities)
+    }
+
+    pub fn cheapest_call_only_with(
+        &self,
+        required_capabilities: &NodeCapabilities,
+    ) -> Result<Arc<EthereumAdapter>, Error> {
+        let adapters = all_cheapest_with(&self.call_only_adapters, required_capabilities);
+
+        match adapters.choose(&mut rand::thread_rng()) {
+            Some(adapter) => Ok(adapter),
+            None => {
+                // We fallback to all adapters if no call only adapters is available to retain backward compatibility
+                self.cheapest_with(required_capabilities)
+            }
+        }
+    }
+
     pub fn remove(&mut self, provider: &str) {
-        self.adapters
+        self.full_adapters
             .retain(|adapter| adapter.adapter.provider() != provider);
     }
 }
 
+fn has_adapter_for<'a>(
```

Comment:
This pattern (firehose + calls adapter) seems to only be necessary for ethereum, could we move this logic to the EthereumAdapterSelector? 

Would it be enough to look for an "eth_call" adapter there? In that case, if there is one we use that, if not we can fallback for an archive, otherwise current firehose configurations would be broken.